portfolio/src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -34,20 +35,23 @@
 @WebServlet("/user")
 public class UserServlet extends HttpServlet {
  
-/**
- * Returns a string or a Login URL depending on the user Login status.
- */
+  /**
+   * Returns a JSON string with the login status and the correspondin login/logout URL.
+   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     UserService userService = UserServiceFactory.getUserService();
-    String json;
+    UserAuthentication user = new UserAuthentication();
     if (userService.isUserLoggedIn()) {
-      json = convertToJsonUsingGson("logged");
+      user.url = userService.createLogoutURL("/comments.html");
```

Comment:
If you're always setting the url to the same place, you can move it outside the if...else... block!

Reply:
If I understood you correctly, the problem is that I need to know the login status in order to create a login or logout URL.

Reply:
Oh no, what I'm saying is you always set user.url = userService.createLogoutURL("/comments.html"); independent of the user login status. So, you could do this:

user.url = userService.createLogoutURL("/comments.html");
user.loginStatus = userService.isUserLoggedIn();

Rather than that bit if...else... block :)

Reply:
I apologize! I didn't notice that 'createLogoutURL' and 'createLoginURL' were different. Either way, your destination URL is the same ("/comments.html"), so you could set that outside the if...else...

You have two options otherwise, I think. Either:

String redirectUrl = "/comments.html" (maybe put a comment here mentioning that the destination URL is the same for both login and logout)
if (userService.isUserLoggedIn()) {
   user.url = userService.createLoginURL(redirectUrl);
} else {
   user.url = userService.createLogoutUrl(redirectUrl);
}

OR you could do this:
String redirectUrl = "/comments.html"
user.url = userService.isUserLoggedIn() ? userService.createLoginURL(redirectUrl) : userService.createLogoutUrl(redirectUrl);

Up to you! The latter option lets you skip the if...else..., but both are fine!

https://stackoverflow.com/questions/10336899/what-is-a-question-mark-and-colon-operator-used-for for reference.

Reply:
Thanks Tim! I think that I'll go with the latter.